src/relay/op/nn/nn.cc

```diff
@@ -14,6 +14,98 @@
 namespace tvm {
 namespace relay {
 
+TVM_REGISTER_NODE_TYPE(DenseAttrs);
+
+
+bool DenseRel(const Array<Type>& types,
+              int num_inputs,
+              const Attrs& attrs,
+              const TypeReporter& reporter) {
+  CHECK_EQ(types.size(), 3);
+  const auto* data = types[0].as<TensorTypeNode>();
+  if (data == nullptr) return false;
+  const auto* weight = types[1].as<TensorTypeNode>();
+  if (weight == nullptr) return false;
+
+  const DenseAttrs* param = attrs.as<DenseAttrs>();
+  CHECK(param != nullptr);
+
+  CHECK(static_cast<int>(data->shape.size()) != 0);
+  Array<tvm::Expr> wshape = weight->shape;
+
+  if (param->units.defined()) {
```

Comment:
if units is defined, we don't need weight to be present, instead we can directly assign to the weight, see the implementation of conv2d

Reply:
Also, need to assign to the weight here, this will infer the shape of the weight if necessary. Please also add a testcase similar to https://github.com/dmlc/tvm/blob/master/tests/python/relay/test_op_level2.py#L14 where weight's shape is not specified but get inferred